### ISSUE TITLE
Framework-agnostic backend

### ISSUE THREAD

**Issue by larsiusprime** [larsiusprime/polymod#6]
I need to refactor this so that Lime/OpenFL isn't assumed, but is simply a pluggable backend. I should create an abstract interface or something that sits between Polymod and Lime, such that it would be possible in the future to add a Heaps, Kha, etc, backend.

**Comment by sh-dave** [comment]
Hi there, i'm currently investigating this as well.

- [polymod](https://github.com/sh-dave/polymod/tree/remove-openfl-deps)
- [polymod-openfl](https://github.com/sh-dave/polymod-openfl) (could maybe work, it's basically all your openfl/lime code, but i haven't actually compiled)
- [polymod-kha](https://github.com/sh-dave/polymod-kha) (not working yet)

Using polymod in this scenario is similar to haxeui V2 and requires the use of 2 libraries: `polymod` (shared core library) + `polymod-XXX` (framework specific code)

What i did so far is:

### core library

- renamed `Polymod` to `PolymodCore` (still lives in in the base repo)
- `Polymod` lives in the framework specific repositories now, and is currently just a typedef to `PolymodCore`, but might get different initialization options if the framework requires/could benefit from it
- created `polymod/fs/StubFileSystem.hx` (returns null/empty for everything)
- created `polymod/fs/SysFileSystem.hx` (old `#if sys` code)
- The core library will do file i/o via PolymodFileSystem now and doesn't depend on cluttering the code with `#if sys` anymore
- created a `PolymodAssets` wrapper to hide framework specific asset handling (requires more thought)

### openfl backend

- created a polymod-openfl repo and moved all openfl/lime dependent code into it
- created a framework specific `PolymodFileSystem` wrapper using `SysFileSystem` for `#if sys` or `StubFileSystem` in other cases

### kha backend

- created a polymod-kha repo and started testing stuff
- created a framework specific `PolymodFileSystem` wrapper as well
  - implemented a new filesystem for kha's debug-html5 target (running via electron) that actually already seems to work

Currently i'm trying to figure out how to inject loaded mod assets into `kha.Assets`.

Let me know what you think!

**Comment by larsiusprime** [comment]
Hey! This is quite promising. I would love to collaborate with you on this.

I kind of like the idea of keeping everything in one library, and relying on conditional compilation in conjunction with dead code elimination to make sure only the right stuff gets pulled in and any unused framework hooks are ignored.

Basically the three things that need abstracting are:
1. Code that overrides the framework's default asset loading mechanism
2. Direct calls to any framework-specific asset system
3. Framework-specific types for assets (Image, BitmapData, Font, etc)

As an example of a previous effort of mine to make a library framework-agnostic, there's crashdumper:
https://github.com/larsiusprime/crashdumper/tree/master/crashdumper/hooks

Crashdumper relies on an `IHookPlatform` interface to abstract the framework dependency:
https://github.com/larsiusprime/crashdumper/blob/master/crashdumper/hooks/IHookPlatform.hx

Here's the OpenFL implementation:
https://github.com/larsiusprime/crashdumper/blob/master/crashdumper/hooks/openfl/HookOpenFL.hx

And I have a utility function for optionally auto-detecting the framework:
https://github.com/larsiusprime/crashdumper/blob/master/crashdumper/hooks/Util.hx#L30

(Never got around to adding new frameworks to Crashdumper, though it theoretically could be done).

--------

Your approach so far looks like more or less what I would do, but if possible I'd like it all to live in one repo; keeping things separated as we work things out for now is fine, though.

Let me revisit my three points: 

> 1. Code that overrides the framework's default asset loading mechanism

One thing to discuss is how Polymod should be used. For instance, it would be much easier to make everything framework agnostic if we just make them replace each and every single framework-specific equivalent "Assets.get()" call in their projects with a PolymodAssets.get() call or whatever. But, that's a huge refactoring pain and one of the appeals in polymod is that you can just drop it into an existing project. In OpenFL this is easy because it has built-in support for overriding your default asset library for just this sort of purpose.  We'll come back to this one in a second.

> 2. Direct calls to any framework-specific asset system

So if we create `PolymodAssets` system, it's not too hard to make that framework agnostic. You create a standard interface filled with stubs and then allow an implementation backend to be attached to it depending on what the user needs to use. 

In the simplest case, require the user to pass in a list of callbacks for every single getX, getY, getZ call! Then when the user requests a piece of text, or some bytes, or an image, we just defer to those. This might be a good starting point for a "no framework" backend.

For OpenFL, we take the existing implementation and make it the "OpenFLAssetsBackend" or whatever.

This however brings up the next issue:

> 3. Framework-specific types for assets (Image, BitmapData, Font, etc)

Polymod requires some knowledge of the asset types. Probably what we're going to have to do here is create our own enumeration list that can be translated to the native asset types of each framework. `String` and `Bytes` are universal across haxe projects, and we'll surely have `getBytes(assetName)` and `getString(assetName)` as workhorse functions, but we'll need to be able to reformat that to what the native framework expects, and we'll need to be able to do some differing logic depending on the asset type.

It might be best to start with 2. and 3. and leave 1. as the final step. Once we know how to provide a generic `PolymodAssets` interface for fetching stuff, and make sure that can line up with any asset system, then we can look into how to use each framework's capabilities to seamlessly allow overriding of the default global asset fetch system, if that is indeed possible. If the framework doesn't support it, the user will have to call `PolymodAssets` directly probably.

**Comment by larsiusprime** [comment]
Probably the best first step is also to just get a good accounting for the differences in all the various framework's asset systems, so we can start with a plan that doesn't paint us into a corner. Could you tell me all about Kha's asset system, for instance?

**Comment by sh-dave** [comment]
1) I'm very much in favor of using the framework specific asset loading on the users side. The `PolymodAssets` wrapper class i created isn't supposed to be used by them, but only as a `core-libary` to `framework-specific` thing. This would be the place where the framework specific asset injection happens. Thatswhy it's implemented in each of the backend repos. So user code would still use the framework's asset management:
  - `openfl.utils.Assets.getImage('img/bee.png')`
  - `kha.Assets.images.get('img/bee.png')`

3) Why does polymod need to know the type of asset, for the text-merging (i haven't really looked too deep into everything the lib has to offer yet)?

### Kha asset handling

As for asset handling in Kha, it's similar to OpenFL, except you don't have the special TEXT type, but only Blobs (= BINARY). The actual format of text assets is something you handle manually in your game/app. So you have:

- Images
- Sounds
- Blobs
- Videos
- Fonts

And all assets are converted during preprocessing into platform specific formats. Take for example sounds. Source files in Kha are always WAV. When you build for html5, they get converted to OGG, MP3, MP4. [Additional info is in the wiki](https://github.com/Kode/Kha/wiki/Managing-Your-Assets)

To access asset you have a number of ways:

- `kha.Assets.images.my_image` - direct access to a preprocessed image reference via a (macro generated) field in the ImageList; the image has to be loaded first, otherwise the reference is null
- `kha.Assets.images.get('my_image') - same as above, but by name string; the image has to be loaded first, otherwise the reference is null
- `kha.Assets.loadImage('my_image', function( img ) { ... }) - load a preprocessed image and get the reference via callback; `kha.Assets.images.my_image` will also return the same reference after this call
- `kha.Assets.loadImageFromPath(`./some-path/my_image.png`, ..., function( img ) {...}) - load an image by path

As for actual interfaces in code, IMO they only make sense for code that executes differently during runtime. Say a generic `Resource` interface that's implemented by `ImageResource`, `SoundResource`, ...
Designing the whole library via interfaces is an unnecessary limitation, but my opinion on this isn't too strong.

**Comment by larsiusprime** [comment]
Yeah, confirmed:
https://github.com/openfl/lime/blob/78e99bf1d92266a06605e32147c9c6ba3a6a03dc/src/lime/utils/AssetCache.hx#L20

Lime's default asset cache behavior is to have three separate typed caches for Images, Audio Buffers, and Fonts -- storing the fully parsed/loaded object in cache rather than just the raw bytes on disk.

In Lime, the act of fetching the asset itself is also when the cache gets set:
https://github.com/openfl/lime/blob/0c96d222869be5251b85ba5c5568da473dcedc4c/src/lime/utils/Assets.hx#L152

And fetching an asset with the generic "BINARY" type signals to skip the cache:
https://github.com/openfl/lime/blob/0c96d222869be5251b85ba5c5568da473dcedc4c/src/lime/utils/Assets.hx#L94

So for frameworks that have this sort of caching behavior, care will have to be taken that overriding it with Polymod's implementation will not cause unintended behavior.

**Comment by larsiusprime** [comment]
Okay! So I'm done refactoring. The OpenFL backend works like it did before, but now it's separated by a layer of abstraction, and you can slot in a new backend there. I've put in stubs for Kha, Heaps, NME, and LIME.

Current plan:
- @sh-dave does the Kha backend?
- I do the HEAPS backend
- I do the LIME and NME backends, based on small modifications from the OpenFL backend

In the end I went with the "keep everything in one repo" approach, and I found that the Backend interface doesn't really need to do much other than expose a few very simple functions. The bulk of the actual work is done entirely in the backend code, where you override your framework's default asset library.

**Comment by sh-dave** [comment]
Yup will do, I'll clean up my sample code (port of the openfl_hscript one) and contribute that as well. Got the scripts working yesterday, and need to work on unloading mods (or the "default" asset library) now. 
Then i'll take a closer look on the new branch.

**Comment by larsiusprime** [comment]
- [x] OpenFL backend done.
- [x] HEAPS backend done.
- [ ] Kha backend WIP by @sh-dave 
- [ ] NME backend needs assistance*
- [ ] Vanilla Lime backend WIP

*NME proved more challenging than I thought, it'll require assistance from someone like @hughsando to help explain to me how the asset system is structured and how to override it.

**Comment by hughsando** [comment]
What is it you are trying to do?  You can set asset data directly to the Assets.info map if this is what you are thinking.

**Comment by larsiusprime** [comment]
@hughsando:

Take a look here:
https://github.com/larsiusprime/polymod/blob/master/polymod/backends/OpenFLBackend.hx

What I do in OpenFL is create a custom `AssetLibrary` that I replace the default one with, that is the first stop for receiving any `Assets.get<Something>(id)` calls.

Then, it does this:
- Look for an asset with that id from the user's mod files (if so, return it)
- If there's no modded version of that file, fall back to the default asset system and return the default file

Basically I need to know what the proper API do this in NME is. I got a little confused trying to sort out asset factories which seem to operate on a per-assettype basis. Does `Assets.info map` let me set one global custom "asset library"?

**Comment by hughsando** [comment]
So if you know the list of mod files at some time (say, startup), the you can call something like "addModFIles()" from the main routine, which would do something like:
```haxe
for( mod in modFiles())
    Assets.info.set(mod.id, new AssetInto( .... ) )
```
If you want it to me more dynamic (ie, after startup), will need to mod the 'Assets.getInfo' function to call out, say by adding an array of "AssetFinders" which would be `String->AssetInfo` functions

**Comment by larsiusprime** [comment]
@hughsando -- Startup is fine, that's the recommended use case.

If I set my own asset info, that's a mapping of an asset id to a new filename on disk, right? I take it that in this situation NME is still ultimately responsible for making the file system calls that actually fetch the asset? That will probably work for my purposes; though it will mean NME is limited to supporting SYS targets (unless there's a way to point NME to a custom virtual filesystem)

**Comment by larsiusprime** [comment]
@hughsando -- the one big issue is text files. With text, we don't just support wholesale file replacement, but allowing mods to append new lines, or inject payloads into target sections of e.g. xml. This requires polymod to be able to fetch the asset and return the bytes for it --- which I think with NME is currently impossible? I can point to a new file, but nme does the loading. We could do it with temp files and path rewrites maybe, but that feels messy. Am I missing something?

**Comment by larsiusprime** [comment]
@hughsando -- okay, just got file replacement working! Works pretty much perfectly, very easy to do, exactly as you described.

My main focus for now is desktop sys target as HTML5 introduces all sorts of security questions, and mobile isn't exactly a hotbed of modding activity as is. If I can get append/merge logic working for text I'll consider that just fine for my purposes.

**Comment by hughsando** [comment]
You can set a custom way of getting the data by using the "byteFactory" in the nme.Assets, which seems to only get triggered if the asset is not a resource.
So maybe if the bytes are a resource (info.isResource) use your existing method.  The bytes will already reside in memory, so I think this limits overhead.
The non-resource path does this:
```haxe
          if (byteFactory.exists(filename))
               data = byteFactory.get(filename)();
```
So, for the polymod overrides (or all of them), iterate over the assets and set
```haxe
   nme.Assets.byteFactory.set( info.path, Polymod.readFile );
    // or, if you need the id:
   nme.Assets.byteFactory.set( info.path, function(filename) return Polymod.readFile(info.id, filename)  ) ;
```
The readFile function should essentially do: `ByteArray.readFile(filename);` and make any mods from there.

**Comment by larsiusprime** [comment]
@hughsando Wonderful, Hugh, that was exactly the ticket!

Here's the final code I wound up using. Seems to work perfectly!

```haxe
for (key in nme.Assets.info.keys())
{
    var info = nme.Assets.info.get(key);
    if(info.type == TEXT){
        if(info.isResource)
        {
            var origText = PolymodAssets.getText(key);
            var newText = polymodLibrary.mergeAndAppendText(key, origText);
            if(origText != newText)
            {
                var byteArray = nme.utils.ByteArray.fromBytes(Bytes.ofString(newText));
                info.setCache(byteArray, true);
                info.isResource = false;
            }
        }
        else
        {
            nme.Assets.byteFactory.set( info.path, function(){
                var bytes = PolymodFileSystem.getFileBytes(key);
                var origText = Std.string(bytes);
                var newText = polymodLibrary.mergeAndAppendText(key, origText);
                if(origText != newText)
                {
                    return nme.utils.ByteArray.fromBytes(Bytes.ofString(newText));
                }
                return nme.utils.ByteArray.fromBytes(Bytes.ofString(origText));
            });
        }
    }
}
```

**Comment by hughsando** [comment]
Love it.  Glad you did not need to change the nme code.

**Comment by larsiusprime** [comment]
@hughsando -- Indeed :) I suspected NME had some sort of facility like this, just needed someone more familiar than me to point out where it was hiding.

**Comment by larsiusprime** [comment]
Just committed the Vanilla Lime backend. That just leaves Kha and CastleDB.

**Comment by MasterEric** [comment]
The actual functionality for interchangable backends is fully implemented, and backends for most frameworks (including HEAPS, Lime, NME, and OpenFL) have already been provided, and functionality to allow for custom backends is in place too.

I think it is more fitting for the remnants of this issue to be split into #57 and #58 for CastleDB and Kha respectively, and close this one.
